### ISSUE TITLE
Refactor API design for BankID to harmonize with GrandID

### ISSUE THREAD

**Issue by PeterOrneholm** [ActiveLogin/ActiveLogin.Authentication#16]
During the development of the GrandID authentication handler a thorough work went into making sure the API, especially the public API, being developer friendly as well ass following common conventions.
 
We aim to refactor the BankID authentication handler before reaching stable 1.0.0 to use the same API design.

This issue will keep track of that work and, before implementing it, will have a suggested API design described here.

**Comment by PeterOrneholm** [comment]
The proposed API design looks like this:

```csharp
services.AddAuthentication()
        .AddBankId(builder => {
        builder
            .UseProdEnvironment()
            .UseBankIdClientCertificateFromAzureKeyVault(options =>
            {
                options.AzureAdClientId = Configuration.GetValue<string>("ActiveLogin:BankId:ClientCertificate:AzureAdClientId");
                options.AzureAdClientSecret = Configuration.GetValue<string>("ActiveLogin:BankId:ClientCertificate:AzureAdClientSecret");
                options.AzureKeyVaultSecretIdentifier = Configuration.GetValue<string>("ActiveLogin:BankId:ClientCertificate:AzureKeyVaultSecretIdentifier");
            })
            .UseBankIdRootCaCertificate(Path.Combine(_environment.ContentRootPath, Configuration.GetValue<string>("ActiveLogin:BankId:CaCertificate:FilePath")))
            .Add("CustomAuthScheme", "CustomDisplayName", options =>
            {
                // ...
            });
        });
```

With shorthand available for the client certificate options:

```csharp
// ...
.UseBankIdClientCertificateFromAzureKeyVault(Configuration.GetSection("ActiveLogin:BankId:ClientCertificate"))
// ...
```

My biggest concern is what to call the `.Add()` method. In GrandId they support three distinct scenarios (`.AddSameDevice()`, `.AddOtherDevice()`, `.AddChooseDevice()`) and native BankID does not work exactly like that. In the future we might be able to achieve the same using policies and device detection, but we don't at the moment.

Therefore we need something that just adds the BankID scheme with options. Possible names as I see it:
* `.Add()`
* `.AddInstance()`
* `.AddScheme()`

Please provide input and vote on a possible name.

**Comment by fredrik-lundin** [comment]
Are there any obvious use-cases where one would like to add more than one scheme?

**Comment by PeterOrneholm** [comment]
Adding some background to why we need / want to redesign:

Today, the API design fools you to believe you can have multiple configurations for things like `ClientCertificate` and `RootCaCertificate` because you add them as part of adding the BankID Scheme. In the end though, the current design only allows for one `BankIdApiClient` and therefore only one `HttpClient` where these configurations apply.

Given that, we wanted to find a design where it's easy to see that you can change these settings and that they apply for the whole application. It also allows for setting the environment in the `.AddBankID()` config rather than outside of it.

**Comment by PeterOrneholm** [comment]
> Are there any obvious use-cases where one would like to add more than one scheme?

Not at the moment, but that is because we have not implemented policies. BankId recomends you to have multiple buttons that enforces different policies in the BankId call. using those policies you can achive oretty much the same scenarios GrandId has (Same/other device). So, once we have support for applying different policies on different buttons we have.

**Comment by fredrik-lundin** [comment]
> > Are there any obvious use-cases where one would like to add more than one scheme?
> 
> Not at the moment, but that is because we have not implemented policies. BankId recomends you to have multiple buttons that enforces different policies in the BankId call. using those policies you can achive oretty much the same scenarios GrandId has (Same/other device). So, once we have support for applying different policies on different buttons we have.

Are there a lot of policies, or a pre-defined set of scenarios? I yes, then we might want to take a similar route as we did for GrandID (e.g. `AddSameDevice`, `AddOtherDevice`), but with fitting methods of course.

**Comment by PeterOrneholm** [comment]
> Are there a lot of policies, or a pre-defined set of scenarios supported? If yes, then we might want to take a similar route as we did for GrandID (e.g. `AddSameDevice`, `AddOtherDevice`), but with fitting methods of course.

There are a few of them, like "Mobile BankID", "BankID on fil"; "Nordea" and so on. They are described under section 14.5 (page 24) here: https://www.bankid.com/assets/bankid/rp/bankid-relying-party-guidelines-v3.2.1.pdf

They can also be combined, so we should have some generic version where you could set whatever combination of policies you like, but also add default setups similar to GrandID, like .AddSameDevice() and .AddOtherDevice().

So, suggestion on what to call the `.Add*()`?

**Comment by PeterOrneholm** [comment]
A separate issue (#18) is now tracking the support for `certificatePolicies`.

**Comment by fredrik-lundin** [comment]
> There are a few of them, like "Mobile BankID", "BankID on fil"; "Nordea" and so on. They are described under section 14.5 (page 24) here: https://www.bankid.com/assets/bankid/rp/bankid-relying-party-guidelines-v3.2.1.pdf
> 
> They can also be combined, so we should have some generic version where you could set whatever combination of policies you like, but also add default setups similar to GrandID, like .AddSameDevice() and .AddOtherDevice().
> 
> So, suggestion on what to call the `.Add*()`?

According to the section in the documentation you pointed to, I see four policies:
* BankID on file
* BankID on smart card
* Mobile BankID
* Nordea e-id on file and on smart card

I suggest adding a method for each of them, just like we did for GrandID (`AddBankIdOnFile` etc).

I don't fully understand what you mean by:
> They can also be combined, so we should have some generic version where you could set whatever combination of policies you like, 

Could you explain in an example of how this could work/be used?

**Comment by PeterOrneholm** [comment]
> I don't fully understand what you mean by:
> 
> > They can also be combined, so we should have some generic version where you could set whatever combination of policies you like,
> 
> Could you explain in an example of how this could work/be used?

Great question! 

So, policies can be combined, with "or", so that multiple auth choices are available. Two common scenarios:

1. You want to login on a desktop, but you don't know if the user has BankID on file or Card, so you combine those two ("BankID on file", "BankID on smart card") so that the user can choose.
2. You only want to have a single button, so you basically add all of the polices and let the user decide on whether he/she uses the phone or BankID on the desktop.

I could, though, see that the four you listed + scenario 1 that I had should be the most common ones. Maybe that's a good start?

**Comment by fredrik-lundin** [comment]
That sounds resonable!

Another idea might be to have another one, `AddWithPolicy`, that takes a policy built with whatever solution #18 ends up with. That could be a nice generic version, which is still _controlled_ to some degree.

**Comment by PeterOrneholm** [comment]
> That sounds resonable!
> 
> Another idea might be to have another one, `AddWithPolicy`, that takes a policy built with whatever solution #18 ends up with. That could be a nice generic version, which is still _controlled_ to some degree.

I like that! Do you think that the policy should end up on the Options, so that you still can modify it yourself? Then I'd say we might call it `AddCustom` instead and don't break the MS convention we followed in GrandId. I'd vote for policy on options.

**Comment by fredrik-lundin** [comment]
Good point. Let's try to keep it as close to that pattern as possible.

**Comment by PeterOrneholm** [comment]
Implemented in #23.
